packages/tools/src/utilities/math/polyline/polylineHelper.ts

```diff
@@ -0,0 +1,201 @@
+import type { Types } from '@cornerstonejs/core';
+// Epsilon for floating point comparisons
+export const EPSILON = 1e-7; // A bit larger than glMatrix.EPSILON for robustness
+
+// --- Basic Vector and Point Math ---
+export function vec2Subtract(
+  out: Types.Point2,
+  a: Types.Point2,
+  b: Types.Point2
+): Types.Point2 {
+  out[0] = a[0] - b[0];
+  out[1] = a[1] - b[1];
+  return out;
+}
+
+export function vec2CrossZ(a: Types.Point2, b: Types.Point2): number {
+  return a[0] * b[1] - a[1] * b[0];
+}
+
+export function distanceSquared(p1: Types.Point2, p2: Types.Point2): number {
+  const dx = p1[0] - p2[0];
+  const dy = p1[1] - p2[1];
+  return dx * dx + dy * dy;
+}
+
+export function pointsAreEqual(p1: Types.Point2, p2: Types.Point2): boolean {
+  return Math.abs(p1[0] - p2[0]) < EPSILON && Math.abs(p1[1] - p2[1]) < EPSILON;
+}
+
+// --- Geometric Primitives (Placeholders/Simplified) ---
+
+/**
+ * Calculates the intersection point of two line segments.
+ * IMPORTANT: This is a simplified version. A robust solution needs to handle
+ * collinear overlaps (which might produce two intersection points or an interval),
+ * parallel lines, and floating point inaccuracies much more carefully.
+ * This function's limitations are a likely source of issues for complex polygons.
+ */
+export function robustSegmentIntersection(
+  p1: Types.Point2,
+  p2: Types.Point2, // Segment 1
+  q1: Types.Point2,
+  q2: Types.Point2 // Segment 2
+): Types.Point2 | null {
+  const r = vec2Subtract([0, 0], p2, p1);
+  const s = vec2Subtract([0, 0], q2, q1);
+  const rxs = vec2CrossZ(r, s);
+  const qmp = vec2Subtract([0, 0], q1, p1);
+  const qmpxr = vec2CrossZ(qmp, r);
+
+  if (Math.abs(rxs) < EPSILON) {
+    // Collinear or parallel
+    // Collinear check: if qmpxr is also zero, they are collinear.
+    // Then, need to check for overlap. This is the hard part not fully handled here.
+    // For true robustness, if collinear and overlapping, you'd return the
+    // two endpoints of the overlap segment. This simple version doesn't.
+    if (Math.abs(qmpxr) < EPSILON) {
+      // Check for overlap for collinear segments
+      const p1OnQ = (t: number) => t >= -EPSILON && t <= 1 + EPSILON;
```

Comment:
This `robustSegmentIntersection` function seems to be running in a loop during subtraction. To avoid this, let's not create new functions here; instead, put them outside.